### PR TITLE
JPA map storage: Authorization services no-downtime store

### LINKED ISSUE #9669: JPA map storage: Authorization services no-downtime store
### Description

Create a JPA map storage implementation for Authorization services area

### PR DESCRIPTION
**Pull request by vramik**
Closes #9669

Sending it as a draft as it'd most likely require rebase once https://github.com/keycloak/keycloak/pull/11831 is merged

### CODE REVIEW

**Review by ahus1 — CHANGES_REQUESTED** [review]
I added some review comments. I'm also run it locally, as in the JPA FAQ from the maven command line. Unfortunately I get the error: 
"Caused by: java.lang.IllegalAccessError: Update to non-static final field org.keycloak.models.map.storage.jpa.authSession.entity.JpaAuthenticationSessionEntity.metadata attempted from a different method ($$_hibernate_write_metadata) than the initializer method <init> 
    at org.keycloak.models.map.storage.jpa.authSession.entity.JpaAuthenticationSessionEntity.$$_hibernate_write_metadata (JpaAuthenticationSessionEntity.java)
"
I need some more time and maybe help to continue the review.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/liquibase/updater/MapJpaLiquibaseUpdaterProvider.java [hunk 1]
```diff
@@ -187,6 +187,7 @@ private List<ChangeSet> getLiquibaseUnrunChangeSets(Liquibase liquibase) throws
     private Liquibase getLiquibase(Class modelType, Connection connection, String defaultSchema) throws LiquibaseException {
         MapLiquibaseConnectionProvider liquibaseProvider = session.getProvider(MapLiquibaseConnectionProvider.class);
         String modelName = ModelEntityUtil.getModelName(modelType);
+        modelName = modelName.contains("authz") ? "authz" : modelName;
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/liquibase/updater/MapJpaLiquibaseUpdaterProvider.java:190]: As far as I understand all authz objects (`permission`, `policy`, `resource`, `resourceServer` and `scope`) will live inside one store. It might be useful to initialize objects in certain order to leverage fact that some object is already existing so that it's possible to create foreign keys.

Looking at it again it seems it'll need some re-design. The reason being is that we cache validated models within `JpaMapStorageProviderFactory.validatedModels` set and currents implementation will visit the validation 5 times. 

As of now I can see two possible ways how to solve it.

1. change  `JpaMapStorageProviderFactory.validatedModels` to `JpaMapStorageProviderFactory.validatedModelNames` where we'd cache names.
2. remove this line and also `jpa-authz-changelog-1.xml` (also probably relocate `jpa-authz-*-changelog-1xml`s) and update the order of MapStorage creation within `MapAuthorizationStoreFactory`:

```
        // order of creating stores matters
        MapStorage resourceServerStore = mapStorageProvider.getStorage(ResourceServer.class);
        MapStorage scopeStore = mapStorageProvider.getStorage(Scope.class);
        MapStorage resourceStore = mapStorageProvider.getStorage(Resource.class);
        MapStorage policyStore = mapStorageProvider.getStorage(Policy.class);
        MapStorage permissionTicketStore = mapStorageProvider.getStorage(PermissionTicket.class);
```

I'm currently in favor of 2. option, WDYT? Maybe another 3. option would be the best.

[the remark this answers is not in the document]
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/liquibase/updater/MapJpaLiquibaseUpdaterProvider.java:190]: As discussed, I'd only make it a bit more robust like this (please note that the `modelName == null` moved up to avoid a NPE). Avoiding ping the DB 5 times could be a future optimization.

```
        if (modelName == null) {
            throw new IllegalStateException("Cannot find changlelog for modelClass " + modelType.getName());
        }
        modelName = modelName.startsWith("authz-") ? "authz" : modelName;
```

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/policy/entity/JpaPolicyEntity.java [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+    }
+
+    @Override
+    public void setLogic(Logic logic) {
+        metadata.setLogic(logic);
+    }
+
+    @Override
+    public Map<String, String> getConfigs() {
+        return config.stream().collect(Collectors.toUnmodifiableMap(JpaPolicyConfigEntity::getName, JpaPolicyConfigEntity::getValue));
```
Comment (ahus1) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/policy/entity/JpaPolicyEntity.java:257]: This API would require JDK11, and for now the pom.xml of this module doesn't allow for JDK11. 

Either use a different API, or update the pom.xml to allow JDK11 (like it is done already with the Quarkus distribution).

**Review by ahus1 — CHANGES_REQUESTED** [review]
@vramik - as discussed, the IllegalAccessError doesn't happen any more after some more full builds; maybe adding the `-Pdistribution` helped. 

I added a review comment about the JDK11 API usage in this module we talked about in the call, please have a look.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/liquibase/updater/MapJpaLiquibaseUpdaterProvider.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/liquibase/updater/MapJpaLiquibaseUpdaterProvider.java]: Thank you, added note about possible future optimization to `JpaMapStorageProviderFactory`

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/policy/entity/JpaPolicyEntity.java (hunk 2 above)
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/policy/entity/JpaPolicyEntity.java]: I've changed an API to be JDK8 compatible to be on a safe site. E.g. here https://github.com/keycloak/keycloak/blob/main/docs/building.md it is stated it requires JDK8 or higher. If we decide we'd like to have new storage modules only JDK11+ we can revisit it, but I'd say in different PR.

**Review by ahus1 — DISMISSED** [review]
I see that the build is green, and a manual test for login worked. I also reviewed all the code changes. Therefore this PR is good to be merged for me. 

I'd love to see at least the model tests run for JPA, but this is something for a different PR that would then take care of all tests that might fail with map JPA.

File: model/map-jpa/src/main/resources/META-INF/persistence.xml [hunk 3]
```diff
@@ -4,6 +4,13 @@
     <!--auth-sessions-->
     <class>org.keycloak.models.map.storage.jpa.authSession.entity.JpaRootAuthenticationSessionEntity</class>
     <class>org.keycloak.models.map.storage.jpa.authSession.entity.JpaAuthenticationSessionEntity</class>
+    <!--authz-->
+    <class>org.keycloak.models.map.storage.jpa.authorization.resourceServer.entity.JpaResourceServerEntity</class>
+    <class>org.keycloak.models.map.storage.jpa.authorization.resource.entity.JpaResourceEntity</class>
+    <class>org.keycloak.models.map.storage.jpa.authorization.resource.entity.JpaResourceAttributeEntity</class>
+    <class>org.keycloak.models.map.storage.jpa.authorization.scope.entity.JpaScopeEntity</class>
+    <class>org.keycloak.models.map.storage.jpa.authorization.permission.entity.JpaPermissionEntity</class>
+    <class>org.keycloak.models.map.storage.jpa.authorization.policy.entity.JpaPolicyEntity</class>
```
[the remark this answers is not in the document]
Comment (vramik) [reply model/map-jpa/src/main/resources/META-INF/persistence.xml:13]: Missed that, thank you. Wondering why it didn't complain in the first place. Maybe it's not used by Hibernate. In that case we might be able to remove the file alltogether.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/policy/entity/JpaPolicyEntity.java [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+
+    @Column(name = "resource_id")
+    @ElementCollection
+    @CollectionTable(name = "kc_authz_policy_resource", joinColumns = @JoinColumn(name = "policy_id", nullable = false))
+    private final Set<UUID> resourceIds = new HashSet<>();
+
+    @Column(name = "associated_policy_id")
+    @ElementCollection
+    @CollectionTable(name = "kc_authz_policy_associated_policy", joinColumns = @JoinColumn(name = "policy_id", nullable = false))
+    private final Set<UUID> policyIds = new HashSet<>();
```
Comment (vramik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/policy/entity/JpaPolicyEntity.java:113]: Maybe, one more thing to consider would to replace `@ElementCollection`s here by `@OneToMany`. As I was reading about it more and it seems hibernate removes all elements and inserts all of them back with each update (addition or removal) [1].

@ahus1, @sguilhen wdyt? Should we leave it as is for now or do you prefer to possibly change it within this PR?

[1] https://thorben-janssen.com/hibernate-tips-elementcollection/

Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/policy/entity/JpaPolicyEntity.java:113]: @vramik - I'd be ok with this for now if the number of entries is small (in the range of 10-100) and would change seldom. 

Maybe one could add a comment to that.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/policy/entity/JpaPolicyEntity.java [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+
+    @Override
+    public Set<String> getAssociatedPolicyIds() {
+        return policyIds.stream().map(UUID::toString).collect(Collectors.toSet());
+    }
+
+    @Override
+    public void addAssociatedPolicyId(String policyId) {
+        UUID policyUuid = StringKeyConverter.UUIDKey.INSTANCE.fromStringSafe(policyId);
+        if (policyUuid != null) policyIds.add(policyUuid);
```
Comment (ahus1) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/policy/entity/JpaPolicyEntity.java:300]: I had a look what `fromStringSafe` does: it will return a null when the `policyId` is not a valid UUID. And then it is not stored to the database. 

As this silently discards data, it is hard to me to see a situation where it is safe to use. Feel free to open a new issue for that.

File: model/map-jpa/src/main/resources/META-INF/authz/policy/jpa-authz-policy-changelog-1.xml [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            <column name="scope_id" type="UUID">
+                <constraints nullable="false" foreignKeyName="authz_policy_scope_scope_fk" references="kc_authz_scope(id)" deleteCascade="true"/>
+            </column>
+        </createTable>
+        <createIndex tableName="kc_authz_policy_scope" indexName="authz_policy_scope_policy_id">
+            <column name="policy_id"/>
+        </createIndex>
+        <createIndex tableName="kc_authz_policy_scope" indexName="authz_policy_scope_scope_id">
+            <column name="scope_id"/>
+        </createIndex>
```
Comment (ahus1) [inline model/map-jpa/src/main/resources/META-INF/authz/policy/jpa-authz-policy-changelog-1.xml:76]: AFAIK policy_id + scope_id combined should be unique in the table, as it is used as a `Set`. Please add a DB constraint for that, with the usually deferrable and initiallyDeferred as Hibernate might change the order of delete and insert.

This might be true for other `ElementCollection` tables as well.

**Review by ahus1 — CHANGES_REQUESTED** [review]
I found case where a DB constraint might possibly be missing, see below.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/policy/entity/JpaPolicyEntity.java (hunk 5 above)
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/policy/entity/JpaPolicyEntity.java:300]: I agree, we should rather throw an exception here. I'll let to throw the original `IllegalArgumentException` here, wdyt?

Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/policy/entity/JpaPolicyEntity.java:300]: @vramik: +1 for the exception.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/policy/entity/JpaPolicyEntity.java (hunk 4 above)
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/policy/entity/JpaPolicyEntity.java:113]: I've checked it with Pedro and it seems the number of objects should be probably between 10 and 30. I'll add a note for future then, thank you.

File: model/map-jpa/src/main/resources/META-INF/authz/policy/jpa-authz-policy-changelog-1.xml (hunk 6 above)
Comment (vramik) [reply model/map-jpa/src/main/resources/META-INF/authz/policy/jpa-authz-policy-changelog-1.xml:76]: Added, thanks

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/resource/delegate/JpaResourceDelegateProvider.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                        break;
+
+                    case SCOPE_IDS:
+                        setDelegateWithAssociation("scopeIds");
+                        break;
+
+                    case URIS:
+                        setDelegateWithAssociation("uris");
+                        break;
+
```
Comment (vramik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/resource/delegate/JpaResourceDelegateProvider.java:68]: After adding this piece, I can see some `ModelIllegalStateException`s ... need to investigate further

File: testsuite/integration-arquillian/tests/base/pom.xml [hunk 8]
```diff
@@ -880,6 +880,7 @@
                                 <keycloak.map.storage.connectionsJpa.user>${keycloak.map.storage.connectionsJpa.user}</keycloak.map.storage.connectionsJpa.user>
                                 <keycloak.map.storage.connectionsJpa.password>${keycloak.map.storage.connectionsJpa.password}</keycloak.map.storage.connectionsJpa.password>
 
+                                <keycloak.authorization.map.storage.provider>jpa</keycloak.authorization.map.storage.provider>
```
Comment (vramik) [inline testsuite/integration-arquillian/tests/base/pom.xml:883]: Assuming https://github.com/keycloak/keycloak/pull/12119 will be merged first

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/resource/delegate/JpaResourceDelegateProvider.java (hunk 7 above)
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authorization/resource/delegate/JpaResourceDelegateProvider.java]: I don't know exact details, but it seems it doesn't work correctly with `ElementCollection`s ... we might be changing those in future to `@OneToMany` so I removed the `ElementCollection` associations from delegates.

**Review by ahus1 — DISMISSED** [review]
I reviewed the latest changes and tested the UI (at least to the point where it triggers the creation of the tables). Thanks for this PR!

**Review by sguilhen — CHANGES_REQUESTED** [review]
As discussed in the last team call, the authorization models, each having their own table mapping, should have their own JPA entity version.

File: model/map-jpa/src/main/resources/META-INF/authz/policy/jpa-authz-policy-changelog-1.xml [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        <!-- this is deferrable and initiallyDeferred as hibernate will first insert new entries and then delete the old by default -->
+        <addUniqueConstraint tableName="kc_authz_policy_scope" columnNames="policy_id, scope_id" deferrable="true" initiallyDeferred="true" />
```
Comment (sguilhen) [inline model/map-jpa/src/main/resources/META-INF/authz/policy/jpa-authz-policy-changelog-1.xml:78]: If https://github.com/keycloak/keycloak/pull/12217 gets merged first, this will have to change (i.e. be moved into its own changeset for postgres only)

Comment (sguilhen) [reply model/map-jpa/src/main/resources/META-INF/authz/policy/jpa-authz-policy-changelog-1.xml:78]: Same thing for all other unique constraints with `deferrable` and `initiallyDeferred` options.

Comment (vramik) [reply model/map-jpa/src/main/resources/META-INF/authz/policy/jpa-authz-policy-changelog-1.xml]: Thanks, the PR was merged. I updated it accordingly.

**Review by sguilhen — APPROVED** [review]
LGTM now, thank you @vramik !

**Review by hmlnarik — APPROVED** [review]
Approving per @sguilhen and @ahus1 's approvals
